shpc/main/registry/remote.py

```diff
@@ -91,44 +74,69 @@ def override_exists(self, tag):
 
 
 class VersionControl(Provider):
-    def __init__(self, *args, **kwargs):
-        self.tag = kwargs.get("tag")
+
+    # The URL is substituted with user, repo
+    library_url_schemes = {
+        "github.com": "https://%s.github.io/%s/library.json",
+        "gitlab.com": "https://%s.gitlab.io/%s/library.json",
+    }
+
+    # The URL is substituted with repo, branch, module_name
+    web_container_url_schemes = {
+        "github.com": "https://github.com/%s/blob/%s/%s/container.yaml",
+        "gitlab.com": "https://gitlab.com/%s/-/blob/%s/%s/container.yaml",
+    }
+
+    # The URL is substituted with repo, branch, module_name
+    raw_container_url_schemes = {
+        "github.com": "https://raw.githubusercontent.com/%s/%s/%s/container.yaml",
+        "gitlab.com": "https://gitlab.com/%s/-/raw/%s/%s/container.yaml",
+    }
+
+    def __init__(self, source, tag=None, subdir=None):
+        if "://" not in source:
+            if os.path.exists(source):
+                raise ValueError(
+                    "VersionControl registry must be a remote path, not a local one."
+                )
+            # Normalise the URL
+            # Heuristics: if there is a @, it's probably ssh
+            if "@" in self.url:
```

Comment:
Just making a suggestion to get the tests passing, and then I can do a more thorough review.

```suggestion
            if "@" in source:
```
For this loop, we are doing a combination of checking source and then setting the self.url. This leads to the double else, which is a bit too nested for my preference. I think I'd like better to set self.url = source off the bat, and then just have one if statement. If we don't hit the if statement, the self.url remains the same.

I haven't reviewed closely yet, but I see what you are trying to do and I like it!

Reply:
Changed